src/Compilers/VisualBasic/Test/Semantic/Binding/SyncLockTests.vb

```diff
@@ -473,5 +474,194 @@ End Module
 </compilation>).VerifyDiagnostics(
             Diagnostic(ERRID.ERR_CaseElseNoSelect, "Case Else"))
         End Sub
+
+        <Fact>
+        Public Sub LockType_InSyncLock()
+            Dim source = "
+Module Program
+    Sub Main()
+        Dim l = New System.Threading.Lock()
+        SyncLock l
+        End SyncLock
+    End Sub
+End Module
+
+Namespace System.Threading
+    Public Class Lock
+    End Class
+End Namespace
+"
+            CreateCompilation(source).AssertTheseDiagnostics(
+"BC42508: A value of type 'System.Threading.Lock' in SyncLock will use likely unintended monitor-based locking. Consider manually calling 'Enter' and 'Exit' methods in a Try/Finally block instead.
```

Comment:
>Consider manually calling 'Enter' and 'Exit' methods in a Try/Finally block instead. [](http://example.com/codeflow?start=112&length=84)

Consider adding a test verifying that following this suggestion is going to compile.

Reply:
It won't compile currently as the Lock type defined in those tests is just a stub. I can define the Enter/Exit methods but seems that such test would not provide much benefit. Better would be to have such test against real net9 runtime in the future.

Reply:
> It won't compile currently as the Lock type defined in those tests is just a stub. I can define the Enter/Exit methods but seems that such test would not provide much benefit. Better would be to have such test against real net9 runtime in the future.

Please add a prototype comment then.

Reply:
> Please add a prototype comment then.

That would block merging the feature until we have net9, not sure we want that? I added a bullet point to the test plan, though.

Reply:
> That would block merging the feature until we have net9, not sure we want that? 

That just means that we will have to open an issue to track completion of the work if we are fine with merging with the gap. The same decision will have to be made for items in the Test Plan. So, in that sense they are all blocking the merge, unless an explicit decision is made to not block on the specific item.